File: phantom-mac-app/Sources/Phantom/Helpers/ClarificationOptionParser.swift
import Foundation

enum ClarificationOptionParser {
    struct ParseResult {
        let displayText: String
        let options: [ClarificationOption]
    }

    private static let choicePrefixes = [
        "did you mean ",
        "do you mean ",
        "are you asking about ",
        "would you like ",
        "should i "
    ]

    private static let connectors = [
        ", or were you asking about ",
        " or were you asking about ",
        ", or did you mean ",
        " or did you mean ",
        ", or "
    ]

    static func parse(_ answer: String) -> ParseResult {
        let text = answer.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return ParseResult(displayText: text, options: []) }
        let listed = extractListedOptions(text)
        if listed.options.count >= 2 { return listed }
        return ParseResult(displayText: listed.displayText, options: inferFromProse(listed.displayText))
    }

    private static func extractListedOptions(_ text: String) -> ParseResult {
        if let split = splitOptionsHeader(text) {
            let options = readListItems(split.list)
            if options.count >= 2 {
                return ParseResult(displayText: split.body, options: options)
            }
        }
        let trailing = extractTrailingList(text)
        return trailing.options.count >= 2 ? trailing : ParseResult(displayText: text, options: [])
    }

    private static func splitOptionsHeader(_ text: String) -> (body: String, list: String)? {
        let pattern = #"([\s\S]*?)\r?\nOptions:\s*\r?\n([\s\S]+)$"#
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]),
              let match = regex.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)),
              let bodyRange = Range(match.range(at: 1), in: text),
              let listRange = Range(match.range(at: 2), in: text) else { return nil }
        return (
            String(text[bodyRange]).trimmingCharacters(in: .whitespacesAndNewlines),
            String(text[listRange])
        )
    }

    private static func extractTrailingList(_ text: String) -> ParseResult {
        var lines = text.replacingOccurrences(of: "\r\n", with: "\n").split(separator: "\n", omittingEmptySubsequences: false).map(String.init)
        var items: [String] = []
        var consumed = 0
        while let line = lines.last {
            let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.isEmpty, items.isEmpty {
                lines.removeLast()
                consumed += 1
                continue
            }
            guard let item = listItem(from: trimmed) else { break }
            items.insert(item, at: 0)
            lines.removeLast()
            consumed += 1
        }
        guard items.count >= 2, consumed < text.split(separator: "\n", omittingEmptySubsequences: false).count else {
            return ParseResult(displayText: text, options: [])
        }
        let body = lines.joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
        return ParseResult(displayText: body.isEmpty ? text : body, options: toOptions(items))
    }

    private static func readListItems(_ list: String) -> [ClarificationOption] {
        toOptions(list.replacingOccurrences(of: "\r\n", with: "\n").split(separator: "\n").compactMap { listItem(from: String($0)) })
    }

    private static func listItem(from line: String) -> String? {
        let pattern = #"^\s*(?:[-*•]|\d+[.)])\s+(.+?)\s*$"#
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: line, range: NSRange(line.startIndex..., in: line)),
              let range = Range(match.range(at: 1), in: line) else { return nil }
        let value = String(line[range]).trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: ".?"))
        return value.isEmpty ? nil : value
    }

    private static func inferFromProse(_ text: String) -> [ClarificationOption] {
        let lowered = text.lowercased()
        var start: String.Index?
        var prefixLength = 0
        for prefix in choicePrefixes {
            if let range = lowered.range(of: prefix, options: [.backwards]) {
                if start == nil || range.lowerBound >= start! {
                    start = range.lowerBound
                    prefixLength = prefix.count
                }
            }
        }
        let remainder: String
        if let start {
            remainder = String(text[text.index(start, offsetBy: prefixLength)...])
        } else {
            remainder = text
        }
        let parts = splitChoices(remainder)
        return parts.count >= 2 ? toOptions(parts) : []
    }

    private static func splitChoices(_ text: String) -> [String] {
        var parts: [String] = []
        var remaining = text.trimmingCharacters(in: .whitespacesAndNewlines)
        while parts.count < 3, !remaining.isEmpty {
            guard let connector = findConnector(remaining) else {
                let last = cleanChoice(remaining)
                if !last.isEmpty { parts.append(last) }
                break
            }
            let head = cleanChoice(String(remaining[..<connector.index]))
            if !head.isEmpty { parts.append(head) }
            remaining = String(remaining[remaining.index(connector.index, offsetBy: connector.length)...])
                .trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return parts
    }

    private static func findConnector(_ text: String) -> (index: String.Index, length: Int)? {
        var depth = 0
        var index = text.startIndex
        while index < text.endIndex {
            let character = text[index]
            if character == "(" {
                depth += 1
            } else if character == ")", depth > 0 {
                depth -= 1
            } else if depth == 0 {
                for connector in connectors {
                    if text[index...].lowercased().hasPrefix(connector.lowercased()) {
                        return (index, connector.count)
                    }
                }
            }
            index = text.index(after: index)
        }
        return nil
    }

    private static func cleanChoice(_ value: String) -> String {
        var trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: ",;—- "))
        while trimmed.hasSuffix("?") || trimmed.hasSuffix(".") {
            trimmed = String(trimmed.dropLast()).trimmingCharacters(in: .whitespacesAndNewlines)
        }
        let prefixes = ["something else, like ", "something else like "]
        for prefix in prefixes where trimmed.lowercased().hasPrefix(prefix) {
            trimmed = String(trimmed.dropFirst(prefix.count)).trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return trimmed
    }

    private static func toOptions(_ items: [String]) -> [ClarificationOption] {
        var options: [ClarificationOption] = []
        var seen = Set<String>()
        for item in items {
            let question = item.trimmingCharacters(in: .whitespacesAndNewlines)
            let key = question.lowercased()
            guard !question.isEmpty, seen.insert(key).inserted else { continue }
            options.append(ClarificationOption(label: makeLabel(question), question: question))
            if options.count == 3 { break }
        }
        return options.count >= 2 ? options : []
    }

    private static func makeLabel(_ question: String) -> String {
        var label = question
        if label.lowercased().hasPrefix("installing "), label.count > 12 {
            let remainder = label.dropFirst(11)
            label = remainder.prefix(1).uppercased() + String(remainder.dropFirst())
        }
        guard label.count > 72 else { return label }
        let end = label.index(label.startIndex, offsetBy: 72)
        let slice = label[..<end]
        if let space = slice.lastIndex(of: " "), label.distance(from: label.startIndex, to: space) >= 24 {
            return String(label[..<space]).trimmingCharacters(in: CharacterSet(charactersIn: ",; ")) + "…"
        }
        return String(slice).trimmingCharacters(in: CharacterSet(charactersIn: ",; ")) + "…"
    }
}

File: phantom-windows-app/Helpers/ClarificationOptionParser.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace SecureOverlay.Helpers
{
    public static class ClarificationOptionParser
    {
        public readonly record struct ParsedOption(string Label, string Question);
        public readonly record struct ParseResult(string DisplayText, IReadOnlyList<ParsedOption> Options);

        private static readonly Regex OptionsHeader = new(
            @"^(?<body>[\s\S]*?)\r?\nOptions:\s*\r?\n(?<list>[\s\S]+)$",
            RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);

        private static readonly Regex ListItem = new(
            @"^\s*(?:[-*•]|\d+[.)])\s+(.+?)\s*$",
            RegexOptions.Multiline | RegexOptions.CultureInvariant);

        private static readonly string[] ChoicePrefixes =
        {
            "did you mean ",
            "do you mean ",
            "are you asking about ",
            "would you like ",
            "should i "
        };

        private static readonly string[] Connectors =
        {
            ", or were you asking about ",
            " or were you asking about ",
            ", or did you mean ",
            " or did you mean ",
            ", or "
        };

        public static ParseResult Parse(string? answer)
        {
            var text = (answer ?? string.Empty).Trim();
            if (string.IsNullOrEmpty(text))
            {
                return new ParseResult(text, Array.Empty<ParsedOption>());
            }

            var listed = ExtractListedOptions(text);
            if (listed.Options.Count >= 2)
            {
                return listed;
            }

            var inferred = InferFromProse(listed.DisplayText);
            return new ParseResult(listed.DisplayText, inferred);
        }

        private static ParseResult ExtractListedOptions(string text)
        {
            var match = OptionsHeader.Match(text);
            if (match.Success)
            {
                var options = ReadListItems(match.Groups["list"].Value);
                if (options.Count >= 2)
                {
                    return new ParseResult(match.Groups["body"].Value.Trim(), options);
                }
            }

            var trailing = ExtractTrailingList(text);
            return trailing.Options.Count >= 2 ? trailing : new ParseResult(text, Array.Empty<ParsedOption>());
        }

        private static ParseResult ExtractTrailingList(string text)
        {
            var lines = text.Replace("\r\n", "\n").Split('\n');
            var items = new List<string>();
            var consumed = 0;
            for (var i = lines.Length - 1; i >= 0; i--)
            {
                var line = lines[i].Trim();
                if (string.IsNullOrEmpty(line) && items.Count == 0)
                {
                    consumed++;
                    continue;
                }

                var itemMatch = ListItem.Match(line);
                if (!itemMatch.Success)
                {
                    break;
                }

                items.Insert(0, itemMatch.Groups[1].Value.Trim());
                consumed++;
            }

            if (items.Count < 2 || consumed >= lines.Length)
            {
                return new ParseResult(text, Array.Empty<ParsedOption>());
            }

            var body = string.Join("\n", lines[..^consumed]).Trim();
            return new ParseResult(string.IsNullOrEmpty(body) ? text : body, ToOptions(items));
        }

        private static IReadOnlyList<ParsedOption> ReadListItems(string list)
        {
            var items = new List<string>();
            foreach (Match match in ListItem.Matches(list))
            {
                var value = match.Groups[1].Value.Trim().TrimEnd('.', '?');
                if (!string.IsNullOrWhiteSpace(value))
                {
                    items.Add(value);
                }
            }

            return ToOptions(items);
        }

        private static IReadOnlyList<ParsedOption> InferFromProse(string text)
        {
            var lowered = text.ToLowerInvariant();
            var start = -1;
            var prefixLength = 0;
            foreach (var prefix in ChoicePrefixes)
            {
                var at = lowered.LastIndexOf(prefix, StringComparison.Ordinal);
                if (at >= 0 && at >= start)
                {
                    start = at;
                    prefixLength = prefix.Length;
                }
            }

            var remainder = start >= 0
                ? text[(start + prefixLength)..]
                : text;
            var parts = SplitChoices(remainder);
            return parts.Count >= 2 ? ToOptions(parts) : Array.Empty<ParsedOption>();
        }

        private static List<string> SplitChoices(string text)
        {
            var parts = new List<string>();
            var remaining = text.Trim();
            while (parts.Count < 3 && remaining.Length > 0)
            {
                var connector = FindConnector(remaining);
                if (connector == null)
                {
                    var last = CleanChoice(remaining);
                    if (!string.IsNullOrEmpty(last))
                    {
                        parts.Add(last);
                    }
                    break;
                }

                var head = CleanChoice(remaining[..connector.Value.Index]);
                if (!string.IsNullOrEmpty(head))
                {
                    parts.Add(head);
                }

                remaining = remaining[(connector.Value.Index + connector.Value.Length)..].Trim();
            }

            return parts;
        }

        private static (int Index, int Length)? FindConnector(string text)
        {
            var depth = 0;
            for (var i = 0; i < text.Length; i++)
            {
                var ch = text[i];
                if (ch == '(')
                {
                    depth++;
                    continue;
                }

                if (ch == ')' && depth > 0)
                {
                    depth--;
                    continue;
                }

                if (depth != 0)
                {
                    continue;
                }

                foreach (var connector in Connectors)
                {
                    if (i + connector.Length <= text.Length
                        && text.AsSpan(i, connector.Length).Equals(connector.AsSpan(), StringComparison.OrdinalIgnoreCase))
                    {
                        return (i, connector.Length);
                    }
                }
            }

            return null;
        }

        private static string CleanChoice(string value)
        {
            var trimmed = value.Trim().Trim(',', ';', '—', '-', ' ');
            while (trimmed.EndsWith('?') || trimmed.EndsWith('.'))
            {
                trimmed = trimmed[..^1].Trim();
            }

            if (trimmed.StartsWith("something else, like ", StringComparison.OrdinalIgnoreCase))
            {
                trimmed = trimmed["something else, like ".Length..].Trim();
            }
            else if (trimmed.StartsWith("something else like ", StringComparison.OrdinalIgnoreCase))
            {
                trimmed = trimmed["something else like ".Length..].Trim();
            }

            return trimmed;
        }

        private static IReadOnlyList<ParsedOption> ToOptions(IReadOnlyList<string> items)
        {
            var options = new List<ParsedOption>();
            var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            foreach (var item in items)
            {
                var question = item.Trim();
                if (question.Length == 0 || !seen.Add(question))
                {
                    continue;
                }

                options.Add(new ParsedOption(MakeLabel(question), question));
                if (options.Count == 3)
                {
                    break;
                }
            }

            return options.Count >= 2 ? options : Array.Empty<ParsedOption>();
        }

        private static string MakeLabel(string question)
        {
            var label = question;
            if (label.StartsWith("installing ", StringComparison.OrdinalIgnoreCase) && label.Length > 12)
            {
                label = char.ToUpperInvariant(label[11]) + label[12..];
            }

            if (label.Length <= 72)
            {
                return label;
            }

            var cut = label.LastIndexOf(' ', 72);
            if (cut < 24)
            {
                cut = 72;
            }

            return label[..cut].TrimEnd(',', ';', ' ') + "…";
        }
    }
}
